src/Engine/Combat/StrikeParams.java

```diff
@@ -0,0 +1,110 @@
+
+  // Stuff inherited for reference from CombatContext
+  public final int battleRange;
+
+  public double baseDamage;
+  public double attackerHP;
+  public double attackPower;
+  public final boolean isCounter;
+
+  public static BattleParams getAttack(final CombatContext ref)
+  {
+    return new BattleParams(
+        new Combatant(ref.attacker, ref.attackerWeapon, ref.attackerX, ref.attackerY),
+        new Combatant(ref.defender, ref.defenderWeapon, ref.defenderX, ref.defenderY),
+        ref.gameMap, ref.battleRange,
+        ref.attacker.model.getDamageRatio(), ref.attacker.getHP(),
+        ref.defender.model.getDefenseRatio(), ref.defenderTerrainStars,
+        false);
+  }
+  public static BattleParams getCounterAttack(final CombatContext ref, double counterHP)
+  {
+    return new BattleParams(
+        new Combatant(ref.defender, ref.defenderWeapon, ref.defenderX, ref.defenderY),
+        new Combatant(ref.attacker, ref.attackerWeapon, ref.attackerX, ref.attackerY),
+        ref.gameMap, ref.battleRange,
+        ref.defender.model.getDamageRatio(), counterHP,
+        ref.attacker.model.getDefenseRatio(), ref.attackerTerrainStars,
+        true);
+  }
+
+  public StrikeParams(
+      Combatant attacker,
+      GameMap map, int battleRange,
+      double attackPower, double attackerHP,
+      double baseDamage,
+      boolean isCounter)
+  {
+    this.attacker = attacker;
+    this.map = map;
+
+    this.battleRange = battleRange;
+    this.attackPower = attackPower;
+    this.isCounter = isCounter;
+    this.baseDamage = baseDamage;
+
+    this.attackerHP = attackerHP;
+
+    // Apply any last-minute adjustments.
+    attacker.body.CO.buffStrike(this);
+  }
+
+  public double calculateDamage()
+  {
+    //    [B*ACO/100+R]*(AHP/10)*[(200-(DCO+DTR*DHP))/100]
+    double overallPower = (baseDamage * attackPower / 100/*+Random factor?*/) * attackerHP / 10;
+    return overallPower / 10; // original formula was % damage, now it must be HP of damage
+  }
+
+  public static class BattleParams extends StrikeParams
+  {
+    public final Combatant defender;
+
+    public double defenderHP;
+    public double defensePower;
+    public double terrainStars;
+
+    public BattleParams(
+        Combatant attacker, Combatant defender,
+        GameMap map, int battleRange,
+        double attackPower, double attackerHP,
+        double defensePower, double terrainStars,
+        boolean isCounter)
+    {
+      super(attacker,
+          map, battleRange,
+          attackPower, attackerHP,
+          (null == attacker.gun)? 0 : attacker.gun.getDamage(defender.body.model),
+          isCounter);
+      this.defender = defender;
+
+      this.defensePower = defensePower;
+      this.terrainStars = terrainStars;
+      defenderHP = defender.body.getHP();
+
+      // Apply any last-minute adjustments.
+      attacker.body.CO.buffAttack(this);
+      defender.body.CO.buffDefense(this);
+    }
+
+    @Override
+    public double calculateDamage()
```

Comment:
I'm not a super-fan of breaking the calculation out over two different functions like this, and I don't think I see the need for using inheritance here. It just feels like it adds complexity for no tangible benefit.

Reply:
I suppose I could just make `StrikeParams` have defense info, and then it's just defaulted to 100 until `BattleParams` fiddles with it?

Reply:
Thinking about it a bit more, I don't really like that solution since it puts all the calculations into a class that doesn't know about what actually is getting hit.
At some point in the future, I'd like to support multiple formulas for determining `overallDefense`.
Then again, you could argue that that calculation could be put in an object under the `StrikeParams`, and the defending CO could just override the default if necessary.